File: app/src/main/java/ru/andrey/cleandictionary/domain/interactor/translation/TranslationInteractor.java
package ru.andrey.cleandictionary.domain.interactor.translation;

import io.reactivex.Single;
import ru.andrey.cleandictionary.domain.model.Translation;

public interface TranslationInteractor {
    /**
     * Returns Single with model which is carried translation
     * <p>
     * Subscribed on Schedulers.io()
     *
     * @param translation model to be translated
     * @return Single of model wuth translation
     */
    Single<Translation> getTranslation(Translation translation);

    /**
     * Returns Single with saved word
     * <p>
     * Subscribed on Schedulers.io()
     *
     * @param translation model to be saved
     * @return Single of saved model
     */
    Single<Translation> saveWord(Translation translation);
}

File: domain/src/main/java/ru/andrey/domain/interactor/TranslationInteractor.kt
package ru.andrey.domain.interactor

import io.reactivex.Scheduler
import io.reactivex.Single
import ru.andrey.domain.model.Language
import ru.andrey.domain.model.Translation
import ru.andrey.domain.repository.TranslationService

class TranslationInteractor(private val ioScheduler: Scheduler,
                            private val translationService: TranslationService) {

    /**
     * Returns Single with model which is carried translation
     *
     *
     * Subscribed on provided scheduler
     *
     * @param word word to be translated,
     * @return Single of model with translation
     */
    fun getTranslation(word: String, from: Language, to: Language): Single<Translation> {
        return translationService
                .getTranslation(word = word, from = from.languageCode, to = to.languageCode)
                .map { Translation(word, it, from, to) }
                .subscribeOn(ioScheduler)
    }
}
